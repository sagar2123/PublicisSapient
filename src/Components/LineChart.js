import React from "react";
import { Chart } from "react-google-charts";

export const LineChart = ({data, xTitle, yTitle}) => {
    return (
        <div className="chart-container">
            <Chart
                height={'400px'}
                chartType="LineChart"
                loader={<div>Loading Chart</div>}
                data={data}
                options={{
                    hAxis: {
                        title: xTitle,
                    },
                    vAxis: {
                        title: yTitle,
                    },
                    backgroundColor: "f6f6f0"
                }}
                rootProps={{ 'data-testid': '1' }}
                />
        </div>
    )
}