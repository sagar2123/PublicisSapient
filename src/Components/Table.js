import React from 'react';
import { getDataFromLocalStorage ,setDataToLocalStorage } from '../Utility/utils';

export const Table = ({data, headers, updateElementAtIndex}) => {

    const createHeaders = () => {
        return headers.map((header, index) => {
            return <th key={header+index} className="text-medium">{header}</th>;
        });
    };

    const updateUpVoteCount = (operation, currentData, index) => {
        let updatedData = JSON.parse(JSON.stringify(currentData));
        if(operation === "inc"){
            updatedData.points = parseInt(updatedData.points) + 1; 
            // updatedData.voted = true;
        } else {
            updatedData.points = parseInt(updatedData.points) - 1; 
            // updatedData.voted = false;
        }

        const checkIfAlreadyExistInLocalStorage = getDataFromLocalStorage(currentData.objectID);
        if(checkIfAlreadyExistInLocalStorage){
            setDataToLocalStorage({...checkIfAlreadyExistInLocalStorage, ...updatedData});
        } else {
            setDataToLocalStorage({...updatedData});
        }
        updateElementAtIndex(index, updatedData);
    };

    const hideElementRow = (currentData, index) => {
        let updatedData = JSON.parse(JSON.stringify(currentData));
        updatedData.hide = true

        const checkIfAlreadyExistInLocalStorage = getDataFromLocalStorage(currentData.objectID);
        if(checkIfAlreadyExistInLocalStorage){
            setDataToLocalStorage({...checkIfAlreadyExistInLocalStorage, ...updatedData});
        } else {
            setDataToLocalStorage({...updatedData});
        }
        updateElementAtIndex(index, updatedData);
    }

    const createRows = () => {
        return data.map((row, index) => {
            return (
                <>
                { !row.hide &&
                    <tr key={row.objectID}>
                        <td className="text-medium align-center">{row.num_comments || "-"}</td>
                        <td className="text-medium align-center">{row.points || "-"}</td>
                        {/* <td className="text-medium align-center">{
                            !row.voted ?
                            <div onClick={() => {updateUpVoteCount("inc", row, index)}} className={"triangle-up-icon cursor-pointer"}></div> : 
                            <div onClick={() => {updateUpVoteCount("dec", row, index)}} className={"triangle-down-icon cursor-pointer"}></div>
                        }</td> */}
                        <td className="text-medium align-center">
                            <div onClick={() => {updateUpVoteCount("inc", row, index)}} className={"triangle-up-icon cursor-pointer"}></div>
                        </td>
                        <td className="text-medium">
                            <a className="news-anchor" href={row.url}>
                                {row.title || "-"} 
                                <span className={"table-elements-non-highlighted"}> by </span> 
                                {row.author || "-"} 
                            </a>
                            <button className={"text-medium float-right hide-button btn-transparent cursor-pointer table-elements-non-highlighted"} onClick={() => {hideElementRow(row, index)}}>hide</button>
                        </td>
                    </tr>
                }
                </>
            );
        });
    }

    return(
        <table>
            <thead>
                <tr>
                    {createHeaders()}
                </tr>
            </thead>
            <tbody>
                {createRows()}
            </tbody>
        </table>
    )
}

