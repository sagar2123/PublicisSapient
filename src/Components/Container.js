import React, { useEffect, useState } from 'react';
import { getPageData } from "../Service/apiFactory";
import { getDataFromLocalStorage } from "../Utility/utils";
import { LineChart } from "./LineChart";
import { Table } from './Table';

export const Container =  (props) => {
    const [data, setData] = useState([]);
    
    useEffect(() => {
        loadData();
        if(!props.match.params.pageId){
            props.history.push(`/page/1`);
        }
    }, [props.match.params.pageId]);
    
    console.log("client params", props.match)
    const loadData = async () => {
        const result = await getPageData(props.match.params.pageId);
        const currentData = result.data.hits.map((currentObj) => {
            const currentDataInStorage = getDataFromLocalStorage(currentObj.objectID);
            const currentDataObj = {
                points: currentObj.points,
                title: currentObj.title,
                author: currentObj.author,
                objectID: currentObj.objectID,
                url: currentObj.url,
                num_comments: currentObj.num_comments,
                voted: false
            }
            return currentDataInStorage ? {
                ...currentDataObj,
                ...currentDataInStorage
            } : {
                ...currentDataObj
            }
        });
        setData(currentData);
    };
    

    const updateElementAtParticularIndex = (index, currentIndexData) => {
        let updatedData = JSON.parse(JSON.stringify(data));
        updatedData[index] = {...updatedData[index], ...currentIndexData};
        setData(updatedData);
    }

    const changePage = (operation) => {
        if(operation === "inc"){
            props.history.push(`/page/${parseInt(props.match.params.pageId) + 1}`);
        } else {
            props.history.push(props.match.params.pageId <= 1 ? `/page/1` : `/page/${parseInt(props.match.params.pageId) - 1}`);
        }
    };

    const prepareLineChartData = (data) => {
        let finalData = data.map((dataObj) => {
            return [dataObj.objectID, dataObj.points]
        });
        return [["Id", "Votes"], ...finalData];
    };

    const hideElement = (index) => {
        let updatedData = JSON.parse(JSON.stringify(data));
        updatedData.splice(index, 1);
        setData(updatedData);
    }

    return (
        <div className="container">
            <>
                <Table headers={["Comments", "Vote Count", "Up Vote", "News Details"]} updateElementAtIndex={updateElementAtParticularIndex} hideElement={hideElement} data={data}/>
                <div className="page-controls background-primary">
                    <div className="app-primary-color">
                        <button className="btn-transparent app-primary-color cursor-pointer" onClick={() => {changePage("dec")}}>Previous</button>
                        | 
                        <button className="btn-transparent app-primary-color cursor-pointer" onClick={() => {changePage("inc")}}>Next</button>
                    </div>
                </div>
                <LineChart data={data.length > 0 ? prepareLineChartData(data) : []} xTitle={"Id"} yTitle={"Votes"}/>
            </> 
        </div>
    )
}