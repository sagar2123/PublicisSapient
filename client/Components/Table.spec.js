import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import  { Table }  from "../../src/Components/Table";

const defaultProps = {
	data: [{
		points: 123,
		title: "news data 1",
		author: "random",
		objectID: 1234,
		url: "ab.com",
		num_comments: 123,
		voted: false

	}],
	headers: ["header1", "header2"],
	updateElementAtIndex: () => {},
	hideElement: () => {}
}

describe("<Table />", () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(
			<Table {...defaultProps}/>
		);
	});

	it("renders without crashing and updates the state", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
	});
});
