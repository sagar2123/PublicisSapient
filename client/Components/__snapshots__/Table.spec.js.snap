// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HackerNews /> renders without crashing and updates the state 1`] = `
<table>
  <thead>
    <tr>
      <th
        className="text-medium"
        key="header10"
      >
        header1
      </th>
      <th
        className="text-medium"
        key="header21"
      >
        header2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="1234"
    >
      <td
        className="text-medium align-center"
      >
        123
      </td>
      <td
        className="text-medium align-center"
      >
        123
      </td>
      <td
        className="text-medium align-center"
      >
        <div
          className="triangle-up-icon cursor-pointer"
          onClick={[Function]}
        />
      </td>
      <td
        className="text-medium"
      >
        <a
          className="news-anchor"
          href="ab.com"
        >
          news data 1
          <span
            className="table-elements-non-highlighted"
          >
             by 
          </span>
          random
        </a>
        <button
          className="text-medium float-right hide-button btn-transparent cursor-pointer table-elements-non-highlighted"
          onClick={[Function]}
        >
          hide
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
